harvester/harvester
# Vendor "frozen" wharfie Registry types

Harvester used to maintain and render /etc/rancher/rke2/registries.yaml, and so it relied on the structs in the wharfie/registries package for serializing/deserializing that information. These structs ultimately became the JSON interface between the Harvester dashboard and the settings controller (at least for the containerd-registry setting.)

A newer version of the embedded Rancher took over that responsibility, and now all Harvester has to do is transform the JSON from the frontend and pack it into the RKEConfig in the cluster.provisioning.cattle.io object.

**Problem:**

The wharfie module's registries package recently updated the struct fields with proper JSON tags, but up until this point, the interface that the setting controller shared with the frontend was predicated on those structs and the encoding/json package's default behavior of defaulting to the exported field name's spelling as the JSON key when one was not specified as it was previously.

**Solution:**

Since the wharfie module's only use in Harvester is for the registries package, just vendor those types with their old JSON key representations to avoid breaking the Harvester dashboard, PXE installs, scripts, etc.

**Related Issue:** https://github.com/harvester/harvester/issues/5190 https://github.com/harvester/harvester/issues/5187

**Test plan:**


1. Start on a release prior to 1.3.0-rc2
2. Add a non-default containerd-registry setting:

```console
harv1:~ # kubectl get setting server-version
NAME             VALUE
server-version   v1.2.1
harv1:~ # cat /etc/rancher/rke2/registries.yaml 
{"configs":{},"mirrors":{"cr.example.com":{"endpoint":["https://cr.example.com:5000"]}}}
```

3. Take note of /etc/rancher/rke2/registries.yaml, and the value shown in the Harvester dashboard
4. Upgrade to a release containing this commit
5. Touch the containerd-registry setting (either delete the hash annotation or just add a space somewhere in the containerd-registry value string (but leave it as syntactically correct JSON)
6. Check that /etc/rancher/rke2/registries.yaml and the value shown in the Harvester dashboard are the values prior to the upgrade

```console
harv1:~ # cat /etc/rancher/rke2/registries.yaml 
{"configs":{},"mirrors":{"cr.example.com":{"endpoint":["https://cr.example.com:5000"]}}}
```
